File: JS/Programmers/Level1/04 키패드 누르기.js
function solution(numbers, hand) {
  var answer = "";

  let left_position = 10;
  let right_position = 11;

  let left_index = 0;
  let middle_index = 0;
  let right_index = 0;

  const left = [1, 4, 7, 10];
  const middle = [2, 5, 8, 0];
  const right = [3, 6, 9, 11];

  for (let i = 0; i < numbers.length; i++) {
    if (middle.indexOf(numbers[i]) != -1)
      middle_index = middle.indexOf(numbers[i]);

    if (left.indexOf(left_position) == -1) {
      if (middle.indexOf(left_position) != -1)
        left_index = middle.indexOf(left_position);
    } else {
      left_index = left.indexOf(left_position);
    }

    if (right.indexOf(right_position) == -1) {
      if (middle.indexOf(right_position) != -1)
        right_index = middle.indexOf(right_position);
    } else {
      right_index = right.indexOf(right_position);
    }

    if (left.indexOf(numbers[i]) != -1) {
      left_position = numbers[i];
      answer = answer + "L";
    } else if (right.indexOf(numbers[i]) != -1) {
      right_position = numbers[i];
      answer = answer + "R";
    } else if (
      middle.indexOf(left_position) != -1 &&
      middle.indexOf(right_position) != -1
    ) {
      if (
        Math.abs(middle_index - left_index) ==
        Math.abs(middle_index - right_index)
      ) {
        if (hand == "left") {
          left_position = numbers[i];
          answer = answer + "L";
        } else {
          right_position = numbers[i];
          answer = answer + "R";
        }
      } else if (
        Math.abs(middle_index - left_index) >
        Math.abs(middle_index - right_index)
      ) {
        right_position = numbers[i];
        answer = answer + "R";
      } else {
        left_position = numbers[i];
        answer = answer + "L";
      }
    } else if (
      middle.indexOf(left_position) != -1 &&
      middle.indexOf(right_position) == -1
    ) {
      if (
        Math.abs(middle_index - left_index) ==
        Math.abs(middle_index - right_index) + 1
      ) {
        if (hand == "left") {
          left_position = numbers[i];
          answer = answer + "L";
        } else {
          right_position = numbers[i];
          answer = answer + "R";
        }
      } else if (
        Math.abs(middle_index - left_index) >
        Math.abs(middle_index - right_index) + 1
      ) {
        right_position = numbers[i];
        answer = answer + "R";
      } else {
        left_position = numbers[i];
        answer = answer + "L";
      }
    } else if (
      middle.indexOf(left_position) == -1 &&
      middle.indexOf(right_position) != -1
    ) {
      if (
        Math.abs(middle_index - left_index) + 1 ==
        Math.abs(middle_index - right_index)
      ) {
        if (hand == "left") {
          left_position = numbers[i];
          answer = answer + "L";
        } else {
          right_position = numbers[i];
          answer = answer + "R";
        }
      } else if (
        Math.abs(middle_index - left_index) + 1 >
        Math.abs(middle_index - right_index)
      ) {
        right_position = numbers[i];
        answer = answer + "R";
      } else {
        left_position = numbers[i];
        answer = answer + "L";
      }
    } else {
      if (
        Math.abs(middle_index - left_index) + 1 ==
        Math.abs(middle_index - right_index) + 1
      ) {
        if (hand == "left") {
          left_position = numbers[i];
          answer = answer + "L";
        } else {
          right_position = numbers[i];
          answer = answer + "R";
        }
      } else if (
        Math.abs(middle_index - left_index) + 1 >
        Math.abs(middle_index - right_index) + 1
      ) {
        right_position = numbers[i];
        answer = answer + "R";
      } else {
        left_position = numbers[i];
        answer = answer + "L";
      }
    }
  }
  return answer;
}

/*
[1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5], "right"

L
R 3
L 4
L4 R3 index L1 R0 M1 Length L1 R2
*/

File: C++/Programmers/Level1/04 키패드 누르기.cpp
#include <string>
#include <vector>
#include <stdlib.h>
#include <algorithm>

using namespace std;

string solution(vector<int> numbers, string hand) {
    string answer = "";

    int left_position = 10;
    int right_position = 11;

    int left_index = 0;
    int middle_index = 0;
    int right_index = 0;

    vector<int> left = {1, 4, 7, 10};
    vector<int> middle = {2, 5, 8, 0};
    vector<int> right = {3, 6, 9, 11};

    for(int i=0; i<numbers.size(); i++){
        middle_index = find(middle.begin(), middle.end(), numbers[i]) - middle.begin();
        
        if(find(left.begin(), left.end(), left_position) == left.end()){
            left_index = find(middle.begin(), middle.end(), left_position) - middle.begin();
        }else {
            left_index = find(left.begin(), left.end(), left_position) - left.begin();
        }

        if(find(right.begin(), right.end(), right_position) == right.end()){
            right_index = find(middle.begin(), middle.end(), right_position) - middle.begin();
        }
        else {
            right_index = find(right.begin(), right.end(), right_position) - right.begin();
        }

        // 1, 4, 7
        if(find(left.begin(), left.end(), numbers[i])!=left.end()){
            left_position = numbers[i];
            answer.append("L");
        }
        // 3, 6, 9
        else if(find(right.begin(), right.end(), numbers[i])!=right.end()){
            right_position = numbers[i];
            answer.append("R");
        }
        else if (find(middle.begin(), middle.end(), left_position)!=middle.end() && find(middle.begin(), middle.end(), right_position)!=middle.end()){
            if(abs(middle_index-left_index) == abs(middle_index-right_index)){
                if(hand.compare("left")){
                    right_position = numbers[i];
                    answer.append("R");
                }
                else {
                    left_position = numbers[i];
                    answer.append("L");
                }
            }
            else if (abs(middle_index-left_index) > abs(middle_index-right_index)){
                right_position = numbers[i];
                answer.append("R");
            }
            else {
                left_position = numbers[i];
                answer.append("L");
            }
        }
        else if (find(middle.begin(), middle.end(), left_position)!=middle.end() && find(middle.begin(), middle.end(), right_position)==middle.end()){
            if(abs(middle_index-left_index)==abs(middle_index-right_index)+1){
                if(hand.compare("left")){
                    right_position = numbers[i];
                    answer.append("R");
                }
                else {
                    left_position = numbers[i];
                    answer.append("L");
                }
            }
            else if (abs(middle_index-left_index)>abs(middle_index-right_index)+1){
                right_position = numbers[i];
                answer.append("R");
            }
            else {
                left_position = numbers[i];
                answer.append("L");
            }
        }
        else if (find(middle.begin(), middle.end(), left_position)==middle.end() && find(middle.begin(), middle.end(), right_position)!=middle.end()){
            if(abs(middle_index-left_index)+1==abs(middle_index-right_index)){
                if(hand.compare("left")){
                    right_position = numbers[i];
                    answer.append("R");
                }
                else {
                    left_position = numbers[i];
                    answer.append("L");
                }
            }
            else if (abs(middle_index-left_index)+1>abs(middle_index-right_index)){
                right_position = numbers[i];
                answer.append("R");
            }
            else {
                left_position = numbers[i];
                answer.append("L");
            }
        }
        else {
            if (abs(middle_index-left_index)+1 == abs(middle_index-right_index)+1){
                if(hand.compare("left")){
                    right_position = numbers[i];
                    answer.append("R");
                }
                else if (hand.compare("right")){
                    left_position = numbers[i];
                    answer.append("L");
                }
            }
            else if(abs(middle_index-left_index)+1 > abs(middle_index-right_index)+1){
                right_position = numbers[i];
                answer.append("R");
            }
            else {
                left_position = numbers[i];
                answer.append("L");
            }
        }
    }

    return answer;
}
